boot/bootutil/src/loader.c

```diff
@@ -1975,6 +1978,162 @@ boot_get_slot_usage(struct boot_loader_state *state, uint8_t slot_usage[],
     return image_cnt;
 }
 
+#ifdef MCUBOOT_RAM_LOADING
+
+#if !defined(IMAGE_EXECUTABLE_RAM_START) || !defined(IMAGE_EXECUTABLE_RAM_SIZE)
+#error "Platform MUST define executable RAM bounds in case of RAM_LOADING"
+#endif
+
+/**
+ * Verifies that the image in a slot will be loaded within the predefined bounds
+ * that are allowed to be used by executable images.
+ *
+ * @param img_dst         The address to which the image is going to be copied.
+ * @param img_sz          The size of the image.
+ *
+ * @return                0 on success; nonzero on failure.
+ */
+static int
+boot_verify_ram_loading_address(uint32_t img_dst, uint32_t img_sz)
+{
+    uint32_t img_end_addr;
+
+    if (img_dst < IMAGE_EXECUTABLE_RAM_START) {
+        return BOOT_EBADIMAGE;
+    }
+
+    if (!boot_u32_safe_add(&img_end_addr, img_dst, img_sz)) {
+        return BOOT_EBADIMAGE;
+    }
+
+    if (img_end_addr > (IMAGE_EXECUTABLE_RAM_START +
+                        IMAGE_EXECUTABLE_RAM_SIZE)) {
+        return BOOT_EBADIMAGE;
+    }
+
+    return 0;
+}
+
+/**
+ * Copies an image from a slot in the flash to an SRAM address. The copying
+ * is done sector-by-sector.
+ *
+ * @param  slot     The flash slot of the image to be copied to SRAM.
+ * @param  img_dst  The address at which the image needs to be copied to
+ *                  SRAM.
+ * @param  img_sz   The size of the image that needs to be copied to SRAM.
+ *
+ * @return          0 on success; nonzero on failure.
+ */
+static int
+boot_copy_image_to_sram(int slot, uint32_t img_dst, uint32_t img_sz)
+{
+    int rc;
+    uint32_t bytes_copied = 0;
+    const struct flash_area *fap_src = NULL;
+
+    rc = flash_area_open(flash_area_id_from_image_slot(slot), &fap_src);
+    if (rc != 0) {
+        return BOOT_EFLASH;
+    }
+
+    while (bytes_copied < img_sz) {
```

Comment:
why coping is done sector by sector?
flash_area_read() api allows to get any amount of data (in zephyr).

Reply:
Actually copy is done in BOOT_TMPBUF_SZ size chunks. I think (but uncertain) that it might makes easier to port to any platform and  project. I mean if the the underlying Flash driver is not coming from Zephyr instead a custom driver with own API could have some limitation. I think the overhead is not too big to do the loading in a loop. Let me know if you think this is unnecessary.

Reply:
Yes, I think this is unnecessary - I guess TF-M flash area abstraction driver is not restricted in size., I believe so Mynewt is not as well (as actually zephyr has adopted this abstraction from it, @utzig Am I right?).

Reply:
Yes, it works without the loop on Mynewt as well.

Reply:
ok, going to remove